pkg/domain/infosync/placement_manager.go

```diff
@@ -99,6 +101,94 @@ func (m *mockPlacementManager) GetAllRuleBundles(_ context.Context) ([]*placemen
 	return bundles, nil
 }
 
+type keyRange struct {
+	groupID  string
+	start    string
+	end      string
+	id       string
+	string   int
+	override bool
+	isLeader bool
+}
+
+// CheckBundle check that the rules don't overlap without explicit Override
+// Exported for testing reasons.
+// Tries to match prepareRulesForApply + checkApplyRules from pd.
+// And additionally checks for key overlaps.
+func CheckBundle(bundle *placement.Bundle) error {
+	keys := make([]keyRange, 0, len(bundle.Rules))
+	for _, rule := range bundle.Rules {
+		if rule.GroupID != bundle.ID {
+			// TODO: also check cross IDs
+			continue
+		}
+		keys = append(keys, keyRange{
+			groupID:  rule.GroupID,
+			id:       rule.ID,
+			start:    rule.StartKeyHex,
+			end:      rule.EndKeyHex,
+			override: rule.Override,
+			isLeader: rule.Role == pd.Leader,
+		})
+	}
+	if len(keys) == 0 {
+		return nil
+	}
+	// Skip overridden rules, but only within the bundle, not across groups
+	applyKeys := keys[:0]
+	j := 0
+	for i := 1; i < len(keys); i++ {
+		if keys[i].override {
+			j = i // skip all previous rules in the same group
+			// TODO: Should we only override the matching key range?
+		}
+	}
+	applyKeys = append(applyKeys, keys[j:]...)
+	if len(applyKeys) == 0 {
+		return fmt.Errorf(`ERROR 8243 (HY000): "[PD:placement:ErrBuildRuleList]build rule list failed, no rule left`)
+	}
+
+	// Additionally check for range overlapping leaders.
+	keys = keys[:0]
+	for i := 0; i < len(applyKeys); i++ {
+		if applyKeys[i].isLeader {
+			keys = append(keys, applyKeys[i])
+		}
+	}
+	if len(keys) == 0 {
+		return nil
+	}
+
+	// Sort on Start, id, end
+	// Could use pd's placement.sortRules() instead.
+	sort.Slice(keys, func(i, j int) bool {
+		if keys[i].start == keys[j].start {
+			if keys[i].id == keys[j].id {
+				return keys[i].end < keys[j].end
+			}
+			return keys[i].id < keys[j].id
+		}
+		return keys[i].start < keys[j].start
+	})
```

Comment:
Maybe we can make above logic more simple by:

```
	for _, rule := range bundle.Rules {
		if rule.GroupID != bundle.ID {
			return errors.New(....)
		}
		if rule.Role == pd.Leader && rule.Override {
		        keys = append(keys, keyRange{
			        start:    rule.StartKeyHex,
			        end:      rule.EndKeyHex,
		        })
		}
	}
	if len(keys) == 0 {
		return nil
	}
	// Sort on Start, id, end
	// Could use pd's placement.sortRules() instead.
	sort.Slice(keys, func(i, j int) bool {
		if keys[i].start == keys[j].start {
			return keys[i].end < keys[j].end
		}
		return keys[i].start < keys[j].start
	})
```

Seems [prepareRulesForApply](https://github.com/tikv/pd/blob/master/pkg/schedule/placement/rule_list.go#L102) is handling rules in the same range which requires building actual rules for each range. We do not need to do it for simple. We can just check there is no overlap in non-override leader rules